erts/emulator/beam/erl_bif_binary.c

```diff
@@ -2457,6 +2459,53 @@ static int cleanup_copy_bin_state(Binary *bp)
     return 1;
 }
 
+static BIF_RETTYPE binary_secure_compare(Process *p, Eterm bin1, Eterm bin2)
+{
+    byte *bytes1;
+    byte *bytes2;
+    Uint bitoffs1, bitsize1;
+    Uint bitoffs2, bitsize2;
+    Uint size1, size2, max_size;
+    size_t i;
+    volatile unsigned char acc = 0;
+  
+    if (is_not_binary(bin1) || is_not_binary(bin2))
+        goto bad_arg;
+    
+    ERTS_GET_BINARY_BYTES(bin1, bytes1, bitoffs1, bitsize1); 
+    ERTS_GET_BINARY_BYTES(bin2, bytes2, bitoffs2, bitsize2);
+
+    size1 = binary_size(bin1);
+    size2 = binary_size(bin2);
+
+    max_size = size1 > size2 ? size1 : size2;
+
+    for (i=0; i < max_size; i++) {
+        if (i <= (size1 - 1) && i <= (size2 - 1)) { 
+            acc |= bytes1[i] ^ bytes2[i];
+        } else {
+            acc |= 'a' ^ 'b';
```

Comment:
Added so we always perform an equivalent operation (striving for "near constant" time).

Reply:
This seems dodgy. The compiler will probably optimize a xor of constants out.
How about this:
```
acc |= bytes1[i rem size1] ^ bytes2[i rem size2];
acc |= size1 ^ size2;
```
Or alternately, this:
```
const volatile char A = 'a';
const volatile char B = 'b';
...
acc |= A ^ B;
```

Reply:
A variation on the above may help keep this simpler:

    {   
        volatile char c1 = '\0';
        volatile char c2 = '\0';
        if (i < size1)
            c1 = bytes1[i];
        if (i < size2)
            c2 = bytes2[i];
        acc |= c1 ^ c2;
    }

Reply:
I hate to be a downer but `volatile` reads and writes still go through processor cache so all proposals thus far will leak the size of the secret. Comparing hashes would work fine, but only if the secret is hashed before it can be interacted with and we can't do that behind this interface.

I'd much rather leave all this to a library that does it for a living. Rolling your own crypto is fun, but I wouldn't trust something of my own design and neither should anyone else.

Reply:
@jhogberg The only other alternative is to do it in pure erlang. I will be back off from making changes until I hear either this PR is rejected in full, the C impl is good enough, or we want it in erlang.

Reply:
@jhogberg Your argument about the processor cache isn't relevant.  It is just best-effort to have a constant time compare.  All source code executes using the processor cache.  Here, volatile is used to avoid getting a register.

Reply:
If it is just best effort why go to all this trouble trying to make it look secure.

Reply:
@IngelaAndin Having a constant time comparison is good, but the compiler can always insert malicious source code.  The hardware all generally lacks formal verification and if it was formally verified it is difficult to confirm hardware matches the proof.  Typically we use monolithic kernels for operating systems that lack formal verification.  Erlang/OTP lacks any formal verification.  Why do we bother with development?  To gradually improve on the situation, making the big ball of mud better in a gradual way.

Reply:
I mean that's my take.... it's best effort, there's no way we or anyone can actually get constant time. Not even with ASM AFAIK. I think this function is generally needed and desired.

Reply:
> A variation on the above may help keep this simpler:
> 
> ```
> {   
>     volatile char c1 = '\0';
>     volatile char c2 = '\0';
>     if (i < size1)
>         c1 = bytes1[i];
>     if (i < size2)
>         c2 = bytes2[i];
>     acc |= c1 ^ c2;
> }
> ```

I could be missing something, but switching to this, which is similar to an approach I initially tried returns true when it should not.

```
binary:secure_compare(<<"ttestxx">>, <<"ttestx">>).
true
```

Reply:
> I could be missing something, but switching to this, which is similar to an approach I initially tried returns true when it should not.
> 
> ```
> binary:secure_compare(<<"ttestxx">>, <<"ttestx">>).
> true
> ```
@starbelly I tried it as separate C source code and Erlang source code to show that acc should be 120 for your test when doing this approach:

    #include <stdio.h>
    #include <string.h>
    int main()
    {
        char const * const bytes1 = "ttestxx";
        char const * const bytes2 = "ttestx";
        size_t const size1 = strlen(bytes1);
        size_t const size2 = strlen(bytes2);
        size_t const max_size = (size1 > size2) ? size1 : size2;
        char acc = 0;
        for (size_t i = 0; i < max_size; i++)
        {
            volatile char c1 = '\0';
            volatile char c2 = '\0';
            if (i < size1)
                c1 = bytes1[i];
            if (i < size2)
                c2 = bytes2[i];
            acc |= c1 ^ c2;
        }
        printf("%d\n", acc);
        return 0;
    }



    -module(test).
    -export([compare_constant/2]).
    compare_constant(Test, [_ | _] = Correct) ->
        compare_constant(Test, Correct, 0) =:= 0.
 
    compare_constant([], [], Bits) ->
        io:format("~p~n", [Bits]),
        Bits;
    compare_constant([] = Test, [C2 | Correct], Bits) ->
        compare_constant(Test, Correct, Bits bor (0 bxor C2));
    compare_constant([C1 | Test], [] = Correct, Bits) ->
        compare_constant(Test, Correct, Bits bor (C1 bxor 0));
    compare_constant([C1 | Test], [C2 | Correct], Bits) ->
        compare_constant(Test, Correct, Bits bor (C1 bxor C2)).

    1> l(test).
    {module,test}
    2> test:compare_constant("ttestxx","ttestx").
    120
    false

Reply:
> @jhogberg Your argument about the processor cache isn't relevant. It is just best-effort to have a constant time compare.  All source code executes using the processor cache. Here, volatile is used to avoid getting a register.

Why would I mention it if it isn't relevant, and why do you assume I don't know why they reached for `volatile`? There's no need to patronize me, if you think I'm unclear you should ask me to clarify instead. I'll readily accept that I was, things that seem obvious to me might not be to someone else.

```
acc |= bytes1[i % size1] ^ bytes2[i % size2];
```

Since we start over from the beginning of the shorter one once we're past its size, we'll have a timing difference once it wraps around since it's bound to be hot in cache (having visited it recently). Figuring out the size to the nearest multiple of the cache line size is easy, and with enough prodding you'll be able to figure out the exact size. It'll take time but it's not hard (assuming you control one of the inputs).

The other variants are vulnerable to similar attacks, easier ones at that.

Does knowing the size of the secret matter? I hope not, but it depends entirely on what the user is doing and I don't like advertising something as "secure" when I can't say that with confidence. Getting this right is _difficult_ and I'd much rather leave this to a library that does it for a living.

Reply:
@jhogberg I hear your points and did the first time and I respect them and your expertise in these areas. I guess at this point, I'm just not clear on what to do? Would renaming the function and eliminating the word secure from the doc suffice? Or are you completely opposed?